src/ir.cpp

```diff
@@ -17047,6 +17047,11 @@ static IrInstruction *ir_analyze_fn_call(IrAnalyze *ira, IrInstructionCallSrc *c
                 if (!handle_is_ptr(result_loc->value.type->data.pointer.child_type)) {
                     ir_reset_result(call_instruction->result_loc);
                     result_loc = nullptr;
+                } else {
+                    call_instruction->base.value.type = impl_fn_type_id->return_type;
+                    IrInstruction *casted_value = ir_implicit_cast(ira, &call_instruction->base, result_loc->value.type->data.pointer.child_type);
```

Comment:
Unfortunately, this is not a sound way to solve the problem. This is mixing pass1 and pass2 IR instructions together, which are not meant to be together. It may work in this case, but this is bound to cause even more complexity for future debugging.

I started looking into merging this, or at least solving the two test cases that this PR adds, by solving it in a more general way, with the result location stuff.  21 files changed, 237 insertions, and 241 deletions later, and I've concluded that this is actually blocking on #3728. Reason being that it removes type coercion of array literals to slices. (To fix broken code will require putting a `&` in front of array literals when one wants them to be coerced to slices). This makes solving result location stuff as well as some other comptime bugs a bit easier, since this cast was tricky to support in the compiler. It was the only type coercion that took a reference on behalf of the programmer. My plan is to finish up that branch so that it can be merged, and then start solving some of these result location issues and other recent regressions, along with trying to catch up with merging PRs.

Reply:
I did realize there might be a problem with this when looking into the 2 last examples in #3688 but I figured there might be an easy way to fix that.